File: Module/ALittleIDE/Other/GameLibrary/CEngine/JSScript/CEngine/Utility/System.js
{
if (typeof ALittle === "undefined") ALittle = {};


ALittle.System_GetPlatform = function() {
	return "Web";
}

ALittle.System_GetDeviceID = function() {
	return JavaScript.JSystem_GetDeviceID();
}

ALittle.System_GetLocalIPList = function() {
	return "[]";
}

ALittle.System_IsPhone = function() {
	let user_agent = navigator.userAgent;
	return ALittle.String_Find(user_agent, "Android") !== undefined || ALittle.String_Find(user_agent, "iPhone") !== undefined || ALittle.String_Find(user_agent, "iPad") !== undefined || ALittle.String_Find(user_agent, "iPod") !== undefined;
}

ALittle.System_InstallProgram = function(file_path) {
	ALittle.Error("not support System_InstallProgram");
}

ALittle.System_ClearAIFamily = function() {
	ALittle.Error("not support System_ClearAIFamily");
}

ALittle.System_StartProgram = function(package_name) {
	ALittle.Error("not support System_StartProgram");
}

ALittle.System_BackProgram = function() {
	ALittle.Error("not support System_BackProgram");
}

ALittle.System_GetScreenWidth = function() {
	return window.innerWidth;
}

ALittle.System_GetScreenHeight = function() {
	return window.innerHeight;
}

ALittle.System_GetStatusBarHeight = function() {
	return 0;
}

ALittle.System_ForceExit = function() {
	ALittle.Error("not support System_ForceExit");
}

ALittle.System_Exit = function() {
	ALittle.Error("not support System_Exit");
}

ALittle.System_Restart = function() {
	ALittle.Error("not support System_Restart");
}

ALittle.System_SetFPS = function(fps) {
	ALittle.Error("not support System_SetFPS");
}

ALittle.System_SetAppPauseInterval = function(interval) {
	__CPPAPI_ScheduleSystem.SetAppPauseInterval(interval);
}

ALittle.System_GetAppPauseInterval = function() {
	return __CPPAPI_ScheduleSystem.GetAppPauseInterval();
}

ALittle.System_StartServerSystem = function(core_path, modules) {
	let module_list = [];
	let ___OBJECT_1 = modules;
	for (let name in ___OBJECT_1) {
		let path = ___OBJECT_1[name];
		if (path === undefined) continue;
		ALittle.List_Push(module_list, name + "," + path);
	}
	return __CPPAPI_ScheduleSystem.StartServerSystem(core_path, ALittle.String_Join(module_list, ";"));
}

ALittle.SystemThreadType = {
	FAST : 0,
	MIDDLE : 1,
	SLOW : 2,
}

ALittle.System_SetThreadCount = function(count, thread_type) {
	if (thread_type === undefined) {
		thread_type = ALittle.SystemThreadType.SLOW;
	}
	ALittle.Error("not support System_SetThreadCount");
}

ALittle.System_GetThreadCount = function(thread_type) {
	if (thread_type === undefined) {
		thread_type = ALittle.SystemThreadType.SLOW;
	}
	ALittle.Error("not support System_SetThreadCount");
	return 0;
}

ALittle.System_GetIMESelectList = function() {
	return lua.__CPPAPI_GetIMESelectList();
}

ALittle.System_SetIMEPos = function(x, y) {
	lua.__CPPAPI_SetIMSPos(x, y);
}

ALittle.System_SetIMERect = function(x, y, w, h) {
	lua.__CPPAPI_SetIMSRect(x, y, w, h);
}

ALittle.System_OpenIME = function() {
	lua.__CPPAPI_OpenIME();
}

ALittle.System_CloseIME = function() {
	lua.__CPPAPI_CloseIME();
}

ALittle.System_GetCursorX = function() {
	return lua.__CPPAPI_GetCursorX();
}

ALittle.System_GetCursorY = function() {
	return lua.__CPPAPI_GetCursorY();
}

ALittle.System_SetEditCursor = function() {
	lua.__CPPAPI_SetEditCursor();
}

ALittle.System_SetNormalCursor = function() {
	lua.__CPPAPI_SetNormalCursor();
}

ALittle.System_SetHandCursor = function() {
	lua.__CPPAPI_SetHandCursor();
}

ALittle.System_SetHDragCursor = function() {
	lua.__CPPAPI_SetHDragCursor();
}

ALittle.System_SetVDragCursor = function() {
	lua.__CPPAPI_SetVDragCursor();
}

ALittle.System_SetHVDragCursor = function() {
	lua.__CPPAPI_SetHVDragCursor();
}

ALittle.System_SetHV2DragCursor = function() {
	lua.__CPPAPI_SetHV2DragCursor();
}

ALittle.System_SetCrossDragCursor = function() {
	lua.__CPPAPI_SetCrossDragCursor();
}

ALittle.System_CreateView = function(title, width, height, flag, scale) {
	return JavaScript.JSystem_CreateView(title, width, height, flag, scale);
}

ALittle.System_SetViewTitle = function(title) {
	return JavaScript.JSystem_SetViewTitle(title);
}

ALittle.System_SetViewIcon = function(path) {
	ALittle.Error("not support System_SetViewIcon");
	return false;
}

ALittle.System_SetViewShape = function(path) {
	ALittle.Error("not support System_SetViewShape");
	return false;
}

ALittle.System_SetViewSize = function(width, height) {
	ALittle.Error("not support System_SetViewSize");
}

ALittle.System_SetMaxViewSize = function() {
	ALittle.Error("not support System_SetMaxViewSize");
}

ALittle.System_SetMinViewSize = function() {
	ALittle.Error("not support System_SetMinViewSize");
}

ALittle.System_SetRestoreViewSize = function() {
	ALittle.Error("not support System_SetRestoreViewSize");
}

ALittle.System_RaiseView = function() {
	ALittle.Error("not support System_RaiseView");
}

ALittle.System_ShowView = function() {
	ALittle.Error("not support System_ShowView");
}

ALittle.System_HideView = function() {
	ALittle.Error("not support System_HideView");
}

ALittle.System_GetViewX = function() {
	ALittle.Error("not support System_GetViewX");
	return 0;
}

ALittle.System_GetViewY = function() {
	ALittle.Error("not support System_GetViewY");
	return 0;
}

ALittle.System_SetViewPosition = function(x, y) {
	ALittle.Error("not support SetViewPosition");
}

ALittle.System_GetMaxTextureWidth = function() {
	return 1024;
}

ALittle.System_GetMaxTextureHeight = function() {
	return 1024;
}

ALittle.System_Render = function() {
	return JavaScript.JSystem_Render();
}

ALittle.System_GetClipboardText = function() {
	ALittle.Error("not support System_GetClipboardText");
	return "";
}

ALittle.System_SetClipboardText = function(content) {
	ALittle.Error("not support System_SetClipboardText");
}

ALittle.System_HasClipboardText = function() {
	ALittle.Error("not support System_HasClipboardText");
	return false;
}

ALittle.System_GetClipboardImage = function() {
	ALittle.Error("not support System_GetClipboardImage");
	return undefined;
}

ALittle.System_SetClipboardImage = function(surface) {
	ALittle.Error("not support System_SetClipboardImage");
}

ALittle.System_HasClipboardImage = function() {
	ALittle.Error("not support System_HasClipboardImage");
	return false;
}

ALittle.SystemOrientationType = {
	SDL_ORIENTATION_UNKNOWN : 0,
	SDL_ORIENTATION_LANDSCAPE : 1,
	SDL_ORIENTATION_LANDSCAPE_FLIPPED : 2,
	SDL_ORIENTATION_PORTRAIT : 3,
	SDL_ORIENTATION_PORTRAIT_FLIPPED : 4,
}

ALittle.System_GetDisplayOrientation = function() {
	let match = matchMedia("(orientation: portrait)");
	if (match !== undefined && match.matches) {
		return ALittle.SystemOrientationType.SDL_ORIENTATION_PORTRAIT;
	}
	return ALittle.SystemOrientationType.SDL_ORIENTATION_LANDSCAPE;
}

ALittle.System_EnableScreenSaver = function() {
	ALittle.Error("not support System_EnableScreenSaver");
}

ALittle.System_DisableScreenSaver = function() {
	ALittle.Error("not support System_DisableScreenSaver");
}

ALittle.System_OpenUrlBySystemBrowser = function(url) {
	open(url);
}

ALittle.System_Alert = function(msg) {
	alert(msg);
}

ALittle.System_LoadSurface = function(path) {
	ALittle.Error("not support System_LoadSurface");
	return undefined;
}

ALittle.System_CreateSurface = function(width, height) {
	ALittle.Error("not support System_CreateSurface");
	return undefined;
}

ALittle.System_SaveSurface = function(surface, path) {
	ALittle.Error("not support System_SaveSurface");
	return false;
}

ALittle.System_FreeSurface = function(surface) {
	ALittle.Error("not support System_FreeSurface");
}

ALittle.System_GetSurfaceWidth = function(surface) {
	ALittle.Error("not support System_GetSurfaceWidth");
	return 0;
}

ALittle.System_GetSurfaceHeight = function(surface) {
	ALittle.Error("not support System_GetSurfaceHeight");
	return 0;
}

ALittle.System_CutBlitSurface = function(new_surface, surface, to, from) {
	ALittle.Error("not support System_CutBlitSurface");
	return false;
}

ALittle.System_SetSurfacePixel = function(surface, x, y, color) {
	ALittle.Error("not support System_SetSurfacePixel");
	return false;
}

ALittle.System_GetSurfacePixel = function(surface, x, y) {
	ALittle.Error("not support System_GetSurfacePixel");
	return 0;
}

ALittle.System_GetPixelAlpha = function(color) {
	ALittle.Error("not support System_GetPixelAlpha");
	return 0;
}

ALittle.System_SetPixelAlpha = function(color, alpha) {
	ALittle.Error("not support System_SetPixelAlpha");
	return 0;
}

ALittle.System_GetSurfaceGrid9 = function(surface, type) {
	ALittle.Error("not support System_GetSurfaceGrid9");
	return 0;
}

ALittle.CreateMsgSender = function(heartbeat, check_heartbeat, callback) {
	return ALittle.NewObject(JavaScript.Template(ALittle.MsgSenderTemplate, "ALittle.MsgSenderTemplate<JavaScript.JMsgInterface, JavaScript.JMessageWriteFactory>", JavaScript.JMsgInterface, JavaScript.JMessageWriteFactory), heartbeat, check_heartbeat, A_JSLoopSystem, callback);
}

ALittle.CreateHttpSender = function(ip, port) {
	return ALittle.NewObject(JavaScript.Template(ALittle.HttpSenderTemplate, "ALittle.HttpSenderTemplate<JavaScript.JHttpInterface>", JavaScript.JHttpInterface), ip, port);
}

ALittle.HttpRequest = function(ip, port, path) {
	return new Promise(function(___COROUTINE, ___) {
		let sender = ALittle.CreateHttpSender(ip, port);
		if (___COROUTINE === undefined) {
			___COROUTINE(["当前不是协程", undefined]); return;
		}
		sender.SendRPC(___COROUTINE, path, undefined);
		return;
		___COROUTINE();
	});
}

ALittle.CreateHttpFileSender = function(ip, port, file_path, start_size, callback) {
	return ALittle.NewObject(JavaScript.Template(ALittle.HttpFileSenderTemplate, "ALittle.HttpFileSenderTemplate<JavaScript.JHttpFileInterface>", JavaScript.JHttpFileInterface), ip, port, file_path, start_size, callback);
}

ALittle.HttpDownloadRequest = function(ip, port, file_path, callback) {
	return new Promise(function(___COROUTINE, ___) {
		let sender = ALittle.CreateHttpFileSender(ip, port, file_path, 0, callback);
		if (___COROUTINE === undefined) {
			___COROUTINE(["当前不是协程", undefined]); return;
		}
		sender.SendDownloadRPC(___COROUTINE, file_path, undefined);
		return;
		___COROUTINE();
	});
}

}

File: ALittle/Module/ALittleIDE/Other/GameLibrary/CEngine/Script/CEngine/Utility/System.lua

module("ALittle", package.seeall)

local ___pairs = pairs
local ___ipairs = ipairs


function System_GetPlatform()
	return __CPPAPI_GetPlatform()
end

function System_GetDeviceID()
	return __CPPAPI_GetDeviceID()
end

function System_GetLocalIPList()
	return __CPPAPI_GetLocalIPList()
end

function System_IsPhone()
	return (__CPPAPI_GetPlatform() ~= "Windows")
end

function System_InstallProgram(file_path)
	__CPPAPI_InstallProgram(file_path)
end

function System_ClearAIFamily()
	__CPPAPI_ClearAIFamily()
end

function System_StartProgram(package_name)
	__CPPAPI_StartProgram(package_name)
end

function System_BackProgram()
	__CPPAPI_BackProgram()
end

function System_GetScreenWidth()
	return __CPPAPI_GetScreenWidth()
end

function System_GetScreenHeight()
	return __CPPAPI_GetScreenHeight()
end

function System_GetStatusBarHeight()
	return __CPPAPI_GetStatusBarHeight()
end

function System_ForceExit()
	__CPPAPI_ScheduleSystem:ForceExit()
end

function System_Exit()
	__CPPAPI_ScheduleSystem:Exit()
end

function System_Restart()
	__CPPAPI_ScheduleSystem:Restart()
end

function System_SetFPS(fps)
	__CPPAPI_ScheduleSystem:SetFPS(fps)
end

function System_SetAppPauseInterval(interval)
	__CPPAPI_ScheduleSystem:SetAppPauseInterval(interval)
end

function System_GetAppPauseInterval()
	return __CPPAPI_ScheduleSystem:GetAppPauseInterval()
end

function System_StartServerSystem(core_path, modules)
	local module_list = {}
	for name, path in ___pairs(modules) do
		List_Push(module_list, name .. "," .. path)
	end
	return __CPPAPI_ScheduleSystem:StartServerSystem(core_path, String_Join(module_list, ";"))
end

SystemThreadType = {
	FAST = 0,
	MIDDLE = 1,
	SLOW = 2,
}

function System_SetThreadCount(count, thread_type)
	if thread_type == nil then
		thread_type = SystemThreadType.SLOW
	end
	__CPPAPI_ThreadSystem:SetThreadCount(count, thread_type)
end

function System_GetThreadCount(thread_type)
	if thread_type == nil then
		thread_type = SystemThreadType.SLOW
	end
	return __CPPAPI_ThreadSystem:GetThreadCount(thread_type)
end

function System_GetIMESelectList()
	return __CPPAPI_GetIMESelectList()
end

function System_SetIMEPos(x, y)
	__CPPAPI_SetIMSPos(x, y)
end

function System_SetIMERect(x, y, w, h)
	__CPPAPI_SetIMSRect(x, y, w, h)
end

function System_OpenIME()
	__CPPAPI_OpenIME()
end

function System_CloseIME()
	__CPPAPI_CloseIME()
end

function System_GetCursorX()
	return __CPPAPI_GetCursorX()
end

function System_GetCursorY()
	return __CPPAPI_GetCursorY()
end

function System_SetEditCursor()
	__CPPAPI_SetEditCursor()
end

function System_SetNormalCursor()
	__CPPAPI_SetNormalCursor()
end

function System_SetHandCursor()
	__CPPAPI_SetHandCursor()
end

function System_SetHDragCursor()
	__CPPAPI_SetHDragCursor()
end

function System_SetVDragCursor()
	__CPPAPI_SetVDragCursor()
end

function System_SetHVDragCursor()
	__CPPAPI_SetHVDragCursor()
end

function System_SetHV2DragCursor()
	__CPPAPI_SetHV2DragCursor()
end

function System_SetCrossDragCursor()
	__CPPAPI_SetCrossDragCursor()
end

function System_CreateView(title, width, height, flag, scale)
	return __CPPAPI_RenderSystem:CreateView(title, width, height, flag, scale)
end

function System_SetViewTitle(title)
	return __CPPAPI_RenderSystem:SetViewTitle(title)
end

function System_SetViewIcon(path)
	return __CPPAPI_RenderSystem:SetViewIcon(path)
end

function System_SetViewShape(path)
	return __CPPAPI_RenderSystem:SetViewShape(path)
end

function System_SetViewSize(width, height)
	__CPPAPI_RenderSystem:SetViewSize(width, height)
end

function System_SetMaxViewSize()
	__CPPAPI_RenderSystem:SetMaxViewSize()
end

function System_SetMinViewSize()
	__CPPAPI_RenderSystem:SetMinViewSize()
end

function System_SetRestoreViewSize()
	__CPPAPI_RenderSystem:SetRestoreViewSize()
end

function System_RaiseView()
	__CPPAPI_RenderSystem:RaiseView()
end

function System_ShowView()
	__CPPAPI_RenderSystem:ShowView()
end

function System_HideView()
	__CPPAPI_RenderSystem:HideView()
end

function System_GetViewX()
	return __CPPAPI_RenderSystem:GetViewX()
end

function System_GetViewY()
	return __CPPAPI_RenderSystem:GetViewY()
end

function System_SetViewPosition(x, y)
	__CPPAPI_RenderSystem:SetViewPosition(x, y)
end

function System_GetMaxTextureWidth()
	return __CPPAPI_RenderSystem:GetMaxTextureWidth()
end

function System_GetMaxTextureHeight()
	return __CPPAPI_RenderSystem:GetMaxTextureHeight()
end

function System_Render()
	return __CPPAPI_RenderSystem:Render()
end

function System_GetClipboardText()
	return __CPPAPI_GetClipboardText()
end

function System_SetClipboardText(content)
	__CPPAPI_SetClipboardText(content)
end

function System_HasClipboardText()
	return __CPPAPI_HasClipboardText()
end

function System_GetClipboardImage()
	return __CPPAPI_GetClipboardImage()
end

function System_SetClipboardImage(surface)
	__CPPAPI_SetClipboardImage(surface)
end

function System_HasClipboardImage()
	return __CPPAPI_HasClipboardImage()
end

SystemOrientationType = {
	SDL_ORIENTATION_UNKNOWN = 0,
	SDL_ORIENTATION_LANDSCAPE = 1,
	SDL_ORIENTATION_LANDSCAPE_FLIPPED = 2,
	SDL_ORIENTATION_PORTRAIT = 3,
	SDL_ORIENTATION_PORTRAIT_FLIPPED = 4,
}

function System_GetDisplayOrientation()
	return __CPPAPI_GetDisplayOrientation()
end

function System_EnableScreenSaver()
	__CPPAPI_EnableScreenSaver()
end

function System_DisableScreenSaver()
	__CPPAPI_DisableScreenSaver()
end

function System_OpenUrlBySystemBrowser(url)
	__CPPAPI_OpenUrlBySystemBrowser(url)
end

function System_Alert(msg)
	__CPPAPI_Alert(msg)
end

function System_LoadSurface(path)
	return __CPPAPI_LoadSurface(path)
end

function System_CreateSurface(width, height)
	return __CPPAPI_CreateSurface(width, height)
end

function System_SaveSurface(surface, path)
	return __CPPAPI_SaveSurface(surface, path)
end

function System_FreeSurface(surface)
	__CPPAPI_FreeSurface(surface)
end

function System_GetSurfaceWidth(surface)
	return __CPPAPI_GetSurfaceWidth(surface)
end

function System_GetSurfaceHeight(surface)
	return __CPPAPI_GetSurfaceHeight(surface)
end

function System_CutBlitSurface(new_surface, surface, to, from)
	return __CPPAPI_CutBlitSurface(new_surface, surface, to, from)
end

function System_SetSurfacePixel(surface, x, y, color)
	return __CPPAPI_SetSurfacePixel(surface, x, y, color)
end

function System_GetSurfacePixel(surface, x, y)
	return __CPPAPI_GetSurfacePixel(surface, x, y)
end

function System_GetPixelAlpha(color)
	return __CPPAPI_GetPixelAlpha(color)
end

function System_SetPixelAlpha(color, alpha)
	return __CPPAPI_SetPixelAlpha(color, alpha)
end

function System_GetSurfaceGrid9(surface, type)
	return __CPPAPI_GetSurfaceGrid9(surface, type)
end

function CreateMsgSender(heartbeat, check_heartbeat, callback)
	return Lua.Template(MsgSenderTemplate, "ALittle.MsgSenderTemplate<lua.__CPPAPIMsgInterface, lua.__CPPAPIMessageWriteFactory>", __CPPAPIMsgInterface, __CPPAPIMessageWriteFactory)(heartbeat, check_heartbeat, A_LuaLoopSystem, callback)
end

function CreateHttpSender(ip, port)
	return Lua.Template(HttpSenderTemplate, "ALittle.HttpSenderTemplate<lua.__CPPAPIHttpInterface>", __CPPAPIHttpInterface)(ip, port)
end

function HttpRequest(ip, port, method)
local ___COROUTINE = coroutine.running()
	local sender = CreateHttpSender(ip, port)
	if ___COROUTINE == nil then
		return "当前不是协程", nil
	end
	sender:SendRPC(___COROUTINE, method, nil)
	return coroutine.yield()
end

function CreateHttpFileSender(ip, port, file_path, start_size, callback)
	return Lua.Template(HttpFileSenderTemplate, "ALittle.HttpFileSenderTemplate<lua.__CPPAPIHttpFileInterface>", __CPPAPIHttpFileInterface)(ip, port, file_path, start_size, callback)
end

function HttpDownloadRequest(ip, port, file_path, method, callback)
local ___COROUTINE = coroutine.running()
	local sender = CreateHttpFileSender(ip, port, file_path, 0, callback)
	if ___COROUTINE == nil then
		return "当前不是协程"
	end
	sender:SendDownloadRPC(___COROUTINE, method, nil)
	return coroutine.yield()
end

